src/edge_op.cpp

```diff
@@ -306,6 +339,124 @@ void Manifold::Impl::CollapseColinearEdges(int firstNewVert) {
+                << std::endl;
+//}
+#endif
+
+      // Check which triangle on this halfedge has two other edges that are
+      // concave
+      // const Halfedge& half = halfedge_[i];
+      size_t concaveEdgesTri1 = 0;
+      const size_t tri1 = i / 3;
+      for (size_t e = 0; e < 3; e++) {
+        const size_t edgeIndex = tri1 * 3 + e;
+        if (i / 3 == edgeIndex) continue;
+        const Halfedge& he = halfedge_[edgeIndex];
+        if (he.pairedHalfedge >= 0) {
+          manifold::vec3 tangent =
+              la::cross((manifold::vec3)faceNormal_[tri1],
+                        (manifold::vec3)vertPos_[he.endVert] -
+                            (manifold::vec3)vertPos_[he.startVert]);
+          double tangentProjection = la::dot(
+              (manifold::vec3)faceNormal_[he.pairedHalfedge / 3], tangent);
+          //  If we've found a pair of reflex triangles, add them to the set
+          if (tangentProjection > 1e-5) {
+            concaveEdgesTri1++;
+          }
+        }
+      }
+      size_t concaveEdgesTri2 = 0;
+      const size_t tri2 = halfedge_[i].pairedHalfedge / 3;
+      for (size_t e = 0; e < 3; e++) {
+        const size_t edgeIndex = tri2 * 3 + e;
+        if (halfedge_[i].pairedHalfedge == edgeIndex) continue;
+        const Halfedge& he = halfedge_[edgeIndex];
+        if (he.pairedHalfedge >= 0) {
+          manifold::vec3 tangent =
+              la::cross((manifold::vec3)faceNormal_[tri2],
+                        (manifold::vec3)vertPos_[he.endVert] -
+                            (manifold::vec3)vertPos_[he.startVert]);
+          double tangentProjection = la::dot(
+              (manifold::vec3)faceNormal_[he.pairedHalfedge / 3], tangent);
+          //  If we've found a pair of reflex triangles, add them to the set
+          if (tangentProjection > 1e-5) {
+            concaveEdgesTri2++;
+          }
+        }
+      }
+
+#ifdef MANIFOLD_DEBUG
+      // if (ManifoldParams().verbose > 0) {
+      std::cout << "  Tri " << tri1 << " has " << concaveEdgesTri1
+                << " concave edges" << std::endl;
+      std::cout << "  Tri " << tri2 << " has " << concaveEdgesTri2
+                << " concave edges" << std::endl;
+//}
+#endif
+
+      if (concaveEdgesTri1 == 2) {
+        // Thicken the triangle by creating a hull from the vertices offset
+        // along the normal
+        vec3 offset = faceNormal_[tri1] * tolerance_ * 0.5;
+        Manifold thickSliver = Manifold::Hull(
+            {vertPos_[halfedge_[tri1 * 3].startVert] + offset,
+             vertPos_[halfedge_[tri1 * 3 + 1].startVert] + offset,
+             vertPos_[halfedge_[tri1 * 3 + 2].startVert] + offset,
+             vertPos_[halfedge_[tri1 * 3].startVert] - offset,
+             vertPos_[halfedge_[tri1 * 3 + 1].startVert] - offset,
+             vertPos_[halfedge_[tri1 * 3 + 2].startVert] - offset});
+
+        //*this = this->Boolean(thickSliver, OpType::Subtract);
+      }
+
+      if (concaveEdgesTri2 == 2) {
```

Comment:
I don't think I'm willing to auto-fix topology like this - it's in the same category as thin webs, where the two sides can easily not be neighbors at all.

Reply:
I respect the principle, but I’d be curious to hear of “edge” cases 😄

Since this routine only gloms onto 180* folding edges, it should only activate when they are neighbors 🧐